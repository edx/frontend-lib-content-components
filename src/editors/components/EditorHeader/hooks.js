import React from 'react';
import * as module from './hooks';

export const hooks = {
  isEditing: () => {
    const [isEditing, setIsEditing] = React.useState(false);
    return {
      isEditing,
      startEditing: () => setIsEditing(true),
      stopEditing: () => setIsEditing(false),
    };
  },
  localTitle: ({
    setBlockTitle,
    stopEditing,
    typeHeader,
  }) => {
    const [localTitle, setLocalTitle] = React.useState(typeHeader);
    return {
      updateTitle: () => {
        setBlockTitle(localTitle);
        stopEditing();
      },
      handleChange: (e) => setLocalTitle(e.target.value),
      localTitle,
    };
  },
  handleKeyDown: ({ stopEditing, editorRef }) => (e) => {
    if (e.key === 'Enter') {
      stopEditing();
    }
    if (e.key === 'Tab' && editorRef) {
      e.preventDefault();
      editorRef.current.focus();
    }
  },
};

/* eslint-disable import/prefer-default-export */
export const localTitleHooks = ({
  editorRef,
  setBlockTitle,
  typeHeader,
}) => {
  const { isEditing, startEditing, stopEditing } = module.hooks.isEditing();
  const { localTitle, handleChange, updateTitle } = module.hooks.localTitle({
    setBlockTitle,
    stopEditing,
    typeHeader,
  });
  return {
    isEditing,
    startEditing,
    stopEditing,

    localTitle,
    updateTitle,
    handleChange,

    inputRef: React.createRef(),
    handleKeyDown: module.hooks.handleKeyDown({ stopEditing, editorRef }),
  };
};
