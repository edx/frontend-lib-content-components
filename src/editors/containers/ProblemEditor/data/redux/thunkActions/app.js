export const initialize = () => (dispatch) => {

  const blockValue = useSelector(selectors.app.blockValue);

  //NOTE: These are to-do methods.
  const data = {...MarkdownParser.getProblem(blockValue.metadata.markdown), getSettingsFromMetaData(blockValue.metadata)}
  dispatch(actions.problem.initialize(data));
};

export default { initialize };
