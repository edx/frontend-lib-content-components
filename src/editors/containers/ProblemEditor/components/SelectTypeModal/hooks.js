import {
  useState,
} from 'react';
import { StrictDict } from '../../utils';

import * as module from './hooks';

export const state = StrictDict({
  selected: (val) => useState(val),
});

export default { state };
