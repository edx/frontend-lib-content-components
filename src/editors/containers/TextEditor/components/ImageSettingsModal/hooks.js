import React from 'react';

import { StrictDict } from '../../../../utils';
import * as module from './hooks';

// Simple wrappers for useState to allow easy mocking for tests.
export const state = {
  dimensions: (val) => React.useState(val),
  locked: (val) => React.useState(val),
  local: (val) => React.useState(val),
  lockInitialized: (val) => React.useState(val),
  altText: (val) => React.useState(val),
  isDecorative: (val) => React.useState(val),
};

export const dimKeys = StrictDict({
  height: 'height',
  width: 'width',
});

/**
 * findGcd(numerator, denominator)
 * Find the greatest common denominator of a ratio or fraction.
 * @param {number} numerator - ratio numerator
 * @param {number} denominator - ratio denominator
 * @return {number} - ratio greatest common denominator
 */
export const findGcd = (a, b) => (b ? findGcd(b, a % b) : a);
const checkEqual = (d1, d2) => (d1.height === d2.height && d1.width === d2.width);

/**
 * getValidDimensions({ dimensions, local, locked })
 * Find valid ending dimensions based on start state, request, and lock state
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after move ({ height, width })
 */
export const getValidDimensions = ({
  dimensions,
  local,
  locked,
}) => {
  const out = {};
  let iter;
  const { minInc } = locked;
  const isMin = dimensions.height === minInc.height;

  const keys = (local.height !== dimensions.height)
    ? { changed: dimKeys.height, other: dimKeys.width }
    : { changed: dimKeys.width, other: dimKeys.height };

  const direction = local[keys.changed] > dimensions[keys.changed] ? 1 : -1;

  // don't move down if already at minimum size
  if (direction < 0 && isMin) { return dimensions; }
  // find closest valid iteration of the changed field
  iter = Math.max(Math.round(local[keys.changed] / minInc[keys.changed]), 1);
  // if closest valid iteration is current iteration, move one iteration in the change direction
  if (iter === (dimensions[keys.changed] / minInc[keys.changed])) { iter += direction; }

  out[keys.changed] = iter * minInc[keys.changed];
  out[keys.other] = out[keys.changed] * (locked[keys.other] / locked[keys.changed]);

  return out;
};

/**
 * newDimensions({ dimensions, local, locked })
 * Returns the local dimensions if unlocked or unchanged, and otherwise returns new valid
 * dimensions.
 * @param {obj} dimensions - current stored dimensions
 * @param {obj} local - local (active) dimensions in the inputs
 * @param {obj} locked - locked dimensions
 * @return {obj} - output dimensions after attempted move ({ height, width })
 */
export const newDimensions = ({ dimensions, local, locked }) => (
  (!locked || checkEqual(local, dimensions))
    ? local
    : module.getValidDimensions({ dimensions, local, locked })
);

/**
 * lockDimensions({ dimensions, lockInitialized, setLocked })
 * Lock dimensions if lock initialized.  Store minimum valid increment on lock so
 * that we don't have re-compute.
 * @param {obj} dimensions - current stored dimensions
 * @param {bool} lockInitialized - has the lock state initialized?
 * @param {func} setLocked - set lock state
 */
export const lockDimensions = ({ dimensions, lockInitialized, setLocked }) => {
  if (!lockInitialized) { return; }

  // find minimum viable increment
  let gcd = findGcd(dimensions.width, dimensions.height);
  if ([dimensions.width, dimensions.height].some(v => !Number.isInteger(v / gcd))) {
    gcd = 1;
  }
  const minInc = { width: dimensions.width / gcd, height: dimensions.height / gcd, gcd };
  setLocked({ ...dimensions, minInc });
};

/**
 * dimensionLockHooks({ dimensions })
 * Returns a set of hooks pertaining to the dimension locks.
 * Locks the dimensions initially, on lock initialization.
 * @param {obj} dimensions - current stored dimensions
 * @return {obj} - dimension lock hooks
 *   {func} initializeLock - enable the lock mechanism
 *   {obj} locked - current locked state
 *   {func} lock - lock the current dimensions
 *   {func} unlock - unlock the dimensions
 */
export const dimensionLockHooks = ({ dimensions }) => {
  const [locked, setLocked] = module.state.locked(null);
  const [lockInitialized, setLockInitialized] = module.state.lockInitialized(null);
  const lock = () => module.lockDimensions({ lockInitialized, dimensions, setLocked });

  React.useEffect(lock, [lockInitialized]);

  return {
    initializeLock: () => setLockInitialized(true),
    locked,
    lock,
    unlock: () => setLocked(null),
  };
};

/**
 * dimensionHooks()
 * Returns an object of dimension-focused react hooks.
 * @return {obj} - dimension hooks
 *   {func} onImgLoad - initializes image dimension fields
 *     @param {object} selection - selected image object with possible override dimensions.
 *     @return {callback} - image load event callback that loads dimensions.
 *   {object} locked - current locked state
 *   {func} lock - lock current dimensions
 *   {func} unlock - unlock dimensions
 *   {object} value - current dimension values
 *   {func} setHeight - set height
 *     @param {string} - new height string
 *   {func} setWidth - set width
 *     @param {string} - new width string
 *   {func} updateDimensions - set dimensions based on state
 */
export const dimensionHooks = () => {
  const [dimensions, setDimensions] = module.state.dimensions(null);
  const [local, setLocal] = module.state.local(null);
  const setAll = (value) => {
    setDimensions(value);
    setLocal(value);
  };
  const {
    initializeLock,
    lock,
    locked,
    unlock,
  } = module.dimensionLockHooks({ dimensions });
  return {
    onImgLoad: (selection) => ({ target: img }) => {
      setAll({
        height: selection.height || img.naturalHeight,
        width: selection.width || img.naturalWidth,
      });
      initializeLock();
    },
    locked,
    lock,
    unlock,
    value: local,
    setHeight: (height) => setLocal({ ...local, height: parseInt(height, 10) }),
    setWidth: (width) => setLocal({ ...local, width: parseInt(width, 10) }),
    updateDimensions: () => setAll(module.newDimensions({ dimensions, local, locked })),
  };
};

/**
 * altTextHooks(savedText)
 * Returns a set of react hooks focused around alt text
 * @return {obj} - alt text hooks
 *   {string} value - alt text value
 *   {func} setValue - set alt test value
 *     @param {string} - new alt text
 *   {bool} isDecorative - is the image decorative?
 *   {func} setIsDecorative - set isDecorative field
 *     @param {bool} isDecorative
 */
export const altTextHooks = (savedText) => {
  const [value, setValue] = module.state.altText(savedText || '');
  const [isDecorative, setIsDecorative] = module.state.isDecorative(false);
  return {
    value,
    setValue,
    isDecorative,
    setIsDecorative,
  };
};

/**
 * onInputChange(handleValue)
 * Simple event handler forwarding the event target value to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target value.
 */
export const onInputChange = (handleValue) => (e) => handleValue(e.target.value);

/**
 * onCheckboxChange(handleValue)
 * Simple event handler forwarding the event target checked prop to a given callback
 * @param {func} handleValue - event value handler
 * @return {func} - evt callback that will call handleValue with the event target checked prop.
 */
export const onCheckboxChange = (handleValue) => (e) => handleValue(e.target.checked);

/**
 * onSave({ altText, dimensions, isDecorative, saveToEditor })
 * Handle saving the image context to the text editor
 * @param {string} altText - image alt text
 * @param {object} dimension - image dimensions ({ width, height })
 * @param {bool} isDecorative - is the image decorative?
 * @param {func} saveToEditor - save method for submitting image settings.
 */
export const onSaveClick = ({
  altText,
  dimensions,
  isDecorative,
  saveToEditor,
}) => () => saveToEditor({
  altText,
  dimensions,
  isDecorative,
});

/**
 * isSaveDisabled(altText)
 * Returns true the save button should be disabled (altText is missing and not decorative)
 * @param {object} altText - altText hook object
 *   {bool} isDecorative - is the image decorative?
 *   {string} value - alt text value
 * @return {bool} - should the save button be disabled?
 */
export const isSaveDisabled = (altText) => !altText.isDecorative && (altText.value === '');

export default {
  altText: altTextHooks,
  dimensions: dimensionHooks,
  isSaveDisabled,
  onCheckboxChange,
  onInputChange,
  onSaveClick,
};
