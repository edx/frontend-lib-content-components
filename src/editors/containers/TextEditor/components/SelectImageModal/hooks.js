import React from 'react';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  images: (val) => React.useState(val),
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: e => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: key => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => imageList.filter(
  ({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()),
);

export const displayList = ({ sortBy, searchString, images }) => module.filteredList({
  searchString,
  imageList: Object.values(images),
}).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({
  fetchImages,
  setSelection,
  searchSortProps,
}) => {
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);

  React.useEffect(() => {
    fetchImages({ onSuccess: setImages });
  }, []);

  return {
    images,
    // highlight by id
    selectBtnProps: {
      disabled: !highlighted,
      onClick: () => setSelection(images[highlighted]),
    },
    galleryProps: {
      highlighted,
      onHighlightChange: e => setHighlighted(e.target.value),
      displayList: module.displayList({ ...searchSortProps, images }),
    },
  };
};

export const fileInputHooks = ({ uploadImage }) => {
  const ref = React.useRef();
  const click = () => ref.current.click();
  const resetFile = () => { ref.current.value = ''; };
  const addFile = (e) => uploadImage({
    file: e.target.files[0],
    resetFile,
  });
  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ fetchImages, uploadImage, setSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ fetchImages, setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({ uploadImage });
  const { selectBtnProps, galleryProps } = imgList;

  return {
    fileInput,
    galleryProps,
    selectBtnProps,
    searchSortProps,
  };
};

export default {
  imgHooks,
};
