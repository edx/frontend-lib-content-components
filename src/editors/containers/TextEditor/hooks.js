import {
  useRef, useEffect, useCallback, useState,
} from 'react';

import { StrictDict } from '../../utils';
import tinyMCE from '../../data/constants/tinyMCE';
import pluginConfig from './pluginConfig';
import * as appHooks from '../../hooks';
import * as module from './hooks';

export const { nullMethod, navigateCallback, navigateTo } = appHooks;

export const state = StrictDict({
  isModalOpen: (val) => useState(val),
  imageSelection: (val) => useState(val),
  refReady: (val) => useState(val),
});

/**
 * Create A function which takes in a tinyMCE editor instance and adds custom
 * features to the toolbars and UI.
 *
 * @param {openModal} func - a function which opens the custom image modal.
 * @param {setImage} func - a function which sets the current image selected in the image modal.
 * @return {func} - takes in a tinyMCE editor instance and adds custom
 * features to the toolbars and UI.
 */

export const setupCustomBehavior = ({ openModal, setImage }) => (editor) => {
  // image upload button
  editor.ui.registry.addButton(tinyMCE.buttons.imageUploadButton, {
    icon: 'image',
    onAction: openModal,
  });
  // editing an existing image
  editor.ui.registry.addButton(tinyMCE.buttons.editImageSettings, {
    icon: 'image',
    onAction: module.openModalWithSelectedImage({ editor, setImage, openModal }),
  });
  // overriding the code plugin's icon with 'HTML' text
  const openCodeEditor = () => editor.execCommand('mceCodeEditor');
  editor.ui.registry.addButton(tinyMCE.buttons.code, {
    text: 'HTML',
    tooltip: 'Source code',
    onAction: openCodeEditor,
  });
};

/**
 * Setup tinyMCE editor with chosen inputs
 *
 * @param {setEditorRef} func - a function which sets the higher-level reference to the editor.
 * @param {blockValue} object - the xblock object which contains the html.
 * @param {openModal} func - a function which opens the custom image modal.
 * @param {setSelection} func - a function which sets the current image selected in the image modal.
 * @param {initializeEditor} func - a function which tells the higher level data abstraction
 * the editor is ready to be displayed.
 */

export const editorConfig = ({
  setEditorRef,
  blockValue,
  openModal,
  initializeEditor,
  setSelection,
}) => ({
  onInit: (evt, editor) => {
    setEditorRef(editor);
    initializeEditor();
  },
  initialValue: blockValue ? blockValue.data.data : '',
  init: {

    setup: module.setupCustomBehavior({ openModal, setImage: setSelection }),
    plugins: pluginConfig.plugins,
    imagetools_toolbar: pluginConfig.imageToolbar,
    toolbar: pluginConfig.toolbar,
    contextmenu: 'link table',
    ...pluginConfig.config,
    valid_elements: '*[*]',
  },
});

/**
 * A stateful toggle for opening/closing a modal.
 * @return {object} - An with toggle open/closed functions and a state reference.
 *
 * usage:
 *   const {isModalOpen: isOpen, openModal, closeModal} = modalToggle();
 *   openModal();
 *   closeModal();
 */

export const modalToggle = () => {
  const [isOpen, setIsOpen] = module.state.isModalOpen(false);
  return {
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};

/**
 * Creates a function which opens the image modal with the selected image when called.
 *
 * @param {editor} object - a tinyMCE editor
 * @param {openModal} func - a function which opens the custom image modal.
 * @param {setImage} func - a function which sets the current image selected in the image modal.
 *
 * @return {func} - An with toggle open/closed functions and a state reference.
 *
 * usage:
 *   const openModal= openModalWithSelectedImage ({editor: currentMCE, setImage, openModal})
 *   openModal();
 */

export const openModalWithSelectedImage = ({ editor, setImage, openModal }) => () => {
  const imgHTML = editor.selection.getNode();
  setImage({
    externalUrl: imgHTML.src,
    altText: imgHTML.alt,
    width: imgHTML.width,
    height: imgHTML.height,
  });
  openModal();
};

export const prepareEditorRef = () => {
  const editorRef = useRef(null);
  const setEditorRef = useCallback((ref) => {
    editorRef.current = ref;
  }, []);
  const [refReady, setRefReady] = module.state.refReady(false);
  useEffect(() => setRefReady(true), []);
  return { editorRef, refReady, setEditorRef };
};

/**
 * Creates a function which returns the HTML content of the tinyMCE Editor
 *
 * @param {editorRef} object - a reference to the tinyMCE editor instance in the DOM
 * @return {func} - a function which returns the content of the editor.
 */

export const getContent = ({ editorRef }) => () => editorRef.current?.getContent();

export const selectedImage = (val) => {
  const [selection, setSelection] = module.state.imageSelection(val);
  return {
    clearSelection: () => setSelection(null),
    selection,
    setSelection,
  };
};
