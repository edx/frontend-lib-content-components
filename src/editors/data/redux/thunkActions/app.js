import { StrictDict } from '../../../utils';
import * as mockData from '../../constants/mockData';
import { actions } from '..';
import * as requests from './requests';
import * as module from './app';

export const fetchBlock = () => (dispatch) => {
  dispatch(requests.fetchBlock({
    onSuccess: (response) => dispatch(actions.app.setBlockValue(response)),
    onFailure: (e) => dispatch(actions.app.setBlockValue(e)),
  }));
};

export const fetchUnit = () => (dispatch) => {
  dispatch(requests.fetchUnit({
    onSuccess: (response) => dispatch(actions.app.setUnitUrl(response)),
    onFailure: (e) => dispatch(actions.app.setUnitUrl(e)),
  }));
};

/**
 * @param {string} studioEndpointUrl
 * @param {string} blockId
 * @param {string} courseId
 * @param {string} blockType
 */
export const initialize = (data) => (dispatch) => {
  dispatch(actions.app.initialize(data));
  dispatch(module.fetchBlock());
  dispatch(module.fetchUnit());
};

/**
 * @param {func} onSuccess
 */
export const saveBlock = ({ content, returnToUnit }) => (dispatch) => {
  dispatch(actions.app.setBlockContent(content));
  dispatch(requests.saveBlock({
    content,
    onSuccess: (response) => {
      dispatch(actions.app.setSaveResponse(response));
      returnToUnit();
    },
  }));
};

export const fetchImages = ({ onSuccess }) => () => {
  // get images
  onSuccess(mockData.mockImageData);
};

export default StrictDict({
  fetchBlock,
  fetchUnit,
  initialize,
  saveBlock,
  fetchImages,
});
