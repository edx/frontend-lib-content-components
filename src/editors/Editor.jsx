import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import { blockTypes } from './data/constants/app';
import { thunkActions } from './data/redux';

import TextEditor from './containers/TextEditor/TextEditor';
import VideoEditor from './containers/VideoEditor/VideoEditor';
import ProblemEditor from './containers/ProblemEditor/ProblemEditor';
import EditorFooter from './components/EditorFooter';
import EditorHeader from './components/EditorHeader';

import messages from './messages';
import * as hooks from './hooks';

export const supportedEditors = {
  [blockTypes.html]: TextEditor,
  [blockTypes.video]: VideoEditor,
  [blockTypes.problem]: ProblemEditor,
};

export const Editor = ({
  courseId,
  blockType,
  blockId,
  studioEndpointUrl,
  // redux
  initialize,
}) => {
  hooks.initializeApp({
    initialize,
    data: {
      blockId,
      blockType,
      courseId,
      studioEndpointUrl,
    },
  });

  const { editorRef, refReady, setEditorRef } = hooks.prepareEditorRef();

  const EditorComponent = supportedEditors[blockType];

  return (
    <div className="d-flex flex-column vh-100">
      <div
        className="pgn__modal-fullscreen"
        role="dialog"
        aria-label={blockType}
      >
        {refReady && (
          <>
            <EditorHeader editorRef={editorRef} />
            {(EditorComponent !== undefined)
              ? <EditorComponent {...{ setEditorRef }} />
              : <FormattedMessage {...messages.couldNotFindEditor} />}
            <EditorFooter editorRef={editorRef} />
          </>
        )}

      </div>
    </div>
  );
};
Editor.defaultProps = {
  courseId: null,
  blockId: null,
  studioEndpointUrl: null,
};

Editor.propTypes = {
  courseId: PropTypes.string,
  blockType: PropTypes.string.isRequired,
  blockId: PropTypes.string,
  studioEndpointUrl: PropTypes.string,
  // redux
  initialize: PropTypes.func.isRequired,
};
export const mapStateToProps = () => ({});
export const mapDispatchToProps = {
  initialize: thunkActions.app.initialize,
};

export default connect(mapStateToProps, mapDispatchToProps)(Editor);
