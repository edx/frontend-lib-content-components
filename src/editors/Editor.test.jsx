import React from 'react';
import { mount, shallow } from 'enzyme';
import { Editor, mapDispatchToProps, supportedEditors } from './Editor';
import { thunkActions } from './data/redux';
import * as hooks from './hooks';
import { blockTypes } from './data/constants/app';

jest.mock('./hooks', () => ({
  initializeApp: jest.fn(),
  prepareEditorRef: jest.fn().mockName('prepareEditorRef'),
}));

jest.mock('./containers/TextEditor/TextEditor', () => 'TextEditor');
jest.mock('./containers/VideoEditor/VideoEditor', () => 'VideoEditor');
jest.mock('./containers/ProblemEditor/ProblemEditor', () => 'ProblemEditor');
jest.mock('./components/EditorFooter', () => 'EditorFooter');
jest.mock('./components/EditorHeader', () => 'EditorHeader');

const props = {
  courseId: 'course-v1:edX+DemoX+Demo_Course',
  blockId: 'block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4',
  studioEndpointUrl: 'fakeurl.com',
  initialize: jest.fn(),
};

describe('Editor', () => {
  describe('snapshots', () => {
    test('renders no editor when ref isnt ready', () => {
      hooks.prepareEditorRef.mockImplementationOnce(
        () => ({ editorRef: null, refReady: false, setEditorRef: jest.fn() }),
      );
      expect(shallow(<Editor blockType="AblOck" {...props} />)).toMatchSnapshot();
    });
    test.each(Object.values(blockTypes))('renders %p editor when ref is ready', (blockType) => {
      hooks.prepareEditorRef.mockImplementationOnce(
        () => ({ editorRef: { current: 'ref' }, refReady: true, setEditorRef: jest.fn().mockName('setEditorRef') }),
      );
      const wrapper = shallow(<Editor blockType={blockType} {...props} />);
      if(blockType == 'html'){ // snap just one editor to make viewing easier
        expect(wrapper).toMatchSnapshot();
      };
      expect(wrapper.children().children().at(1).is(supportedEditors[blockType])).toBe(true);
    });
    test('presents error message if no relevant editor found and ref ready', () => {
      hooks.prepareEditorRef.mockImplementationOnce(
        () => ({ editorRef: { current: 'ref' }, refReady: true, setEditorRef: jest.fn().mockName('setEditorRef') }),
      );
      expect(shallow(<Editor
        {...props}
        blockType="fAkEBlock"
      />)).toMatchSnapshot();
    });
  });
  describe('mapDispatchToProps', () => {
    test('initialize from thunkActions.app.initialize', () => {
      expect(mapDispatchToProps.initialize).toEqual(thunkActions.app.initialize);
    });
  });
});
