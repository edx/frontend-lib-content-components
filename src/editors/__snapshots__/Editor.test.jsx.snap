// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor snapshots presents error message if no relevant editor found and ref ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="fAkEBlock"
    className="pgn__modal-fullscreen"
    role="dialog"
  >
    <EditorHeader
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
    <FormattedMessage
      defaultMessage="Error: Could Not find Editor"
      description="Error Message Dispayed When An unsopported Editor is desired in V2"
      id="authoring.editorpage.selecteditor.error"
    />
    <EditorFooter
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
  </div>
</div>
`;

exports[`Editor snapshots renders "html" editor when ref is ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="html"
    className="pgn__modal-fullscreen"
    role="dialog"
  >
    <EditorHeader
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
    <TextEditor
      setEditorRef={[MockFunction setEditorRef]}
    />
    <EditorFooter
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
  </div>
</div>
`;

exports[`Editor snapshots renders no editor when ref isnt ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="AblOck"
    className="pgn__modal-fullscreen"
    role="dialog"
  />
</div>
`;
